### DISCUSSION TITLE
modbus device king pigeon S272

### DISCUSSION THREAD

**Discussion by agalindo73** [aldas/modbus-tcp-client#78] [Q&A]
hi aldas
I discovered your repositories related to modbus client, I am interested in knowing if it works for a King Pigeon Modbus RTU TCP 3G / 4G device this device sends a heartbeat every 60 seconds, I want to develop a php server that gets the device's registration states and addresses them to my host where I have the mysql database, this device is configured ip and port.
I would like to explain you better by email if you would like to provide it to me, I am Alonso Galindo from Mexico
regards

**Comment by aldas** [comment]
Looking at http://www.gprs-m2m.com/EnproductShow.asp?ID=423 I would guess this library works with this device.  RTU example can be found here https://github.com/aldas/modbus-tcp-client/blob/master/examples/rtu.php 

If you have problems getting response I suggest checking these issues to help you diagnose what is wrong:
* https://github.com/aldas/modbus-tcp-client/issues/76
* https://github.com/aldas/modbus-tcp-client/issues/71

**Comment by agalindo73** [comment]
Thank you very much for your answer aldas, I was seeing your following code  https://github.com/aldas/go-modbus-client 
Since the rtu device is managed on a 3G / 4G network and these networks handle private IP, so it is difficult to configure the IP on the client, however, this device works with a kiiot platform where I send you the screenshot of the configuration menu You can see that it is configured so that the data reaches the platform of the KIIOT company. However I want to make the code so that my host reaches a personal page where its php code would be implemented in my private host 67.227.237.232 (agalindotelemetria.com.mx) 
![proyectomaestria](https://user-images.githubusercontent.com/73866027/113649258-f6597d80-9642-11eb-8a20-d4752ba8d791.jpg)
and port "to be specified"
because I am an electromechanical engineer but I am not a developer, I just want to make this code functional
You can see in the configuration menu in the part of login messaage that ID was provided to me by the company to access their platform and below is the heartbeat configured as "rep" and "res" that is why it caught my attention a lot code that I put above I think that would be the ideal, but my question is how do I test if I mount it on a host?
![menuconfigurationS272](https://user-images.githubusercontent.com/73866027/113649130-beead100-9642-11eb-930e-6803e7707340.jpg)

**Comment by aldas** [comment]
> You can see that it is configured so that the data reaches the platform of the KIIOT company.

Looking at the screenshot and considering your idea I have more questions than answers.

* That screenshot has fields for login. Modbus protocol does not have security at all. So I assume KIIOT has their own proprietary solution/protocol to achieve security over internet.  Maybe there are some manuals/guides/pdfs that hint how this stuff works. It would be easier if you can link to user guide to that applications it could result in some workarounds.

* There is that "Connect other IOT server" what options are under that?

* You could try running modbus server with https://www.modbusdriver.com/diagslave.html and see if you can configure your gateway to connect to it. It would validate idea for building/setting up your own modbus server to accept write requests periodically

**Comment by agalindo73** [comment]
This is the screenshot that you see is currently configured so that the data goes to the kiiot platform (http://www.my-rtu.com/login) where the King pigeon company provided me with a username and password and also the serial number of the device and it is the login message that you observe in the configuration part, I think that with that serial number that they provided me is to identify my device on the company's platform, I decided to hire their platform, however if not contract the device has the ability to work on a server designed and customized to my needs, for example I can better visualize the process and variables.
As for the part that you observe "connect IoT server", this S272 model does not have that configurable option, the software that they provided me configures several models of the company and the S275 has the IoT option.
But hey, the model that I have S272, they told me that it allows me to connect it to my own server only indicating IP and port, this is where according to what I have investigated is that a web system has to be made with a php code where it is listening to the heartbeat of the device to respond and the device to send the information.
It is here where I was looking for information and I was able to find that you have a modbus client code that could be used. The only thing I want to do is see that your code works with my device but I cannot understand how to implement it in an html page and where I see that a link between the device and the html page
That would be the first step, and later I want the code to send that information to the mysql database

I am attaching the answers of the company's technicians to some emails that I send requesting information

These devices are at a good price for the functionalities that it has and offer the platform at a good price, but I only needed to make a prototype a web server to visualize that information and store it in a database to fulfill a master's thesis.


![kiiotplataform1](https://user-images.githubusercontent.com/73866027/113804363-3daf3f00-9713-11eb-9de0-6e4c81ec74c7.jpg)
![kiiotplataform2](https://user-images.githubusercontent.com/73866027/113804374-4273f300-9713-11eb-8b5a-e2f26d8e19ef.jpg)

![kiiotplataform3](https://user-images.githubusercontent.com/73866027/113804706-f4abba80-9713-11eb-8501-b697f1e37c74.jpg)
![kiiotplataform4](https://user-images.githubusercontent.com/73866027/113804727-ff664f80-9713-11eb-84e9-7da39761ef56.jpg)

**Comment by aldas** [comment]
Okey, assuming your device does not need login "handshake" and can work with regular modbus servers this is what you can do  to begin with

1. Download diagslave utility to see if your gateway can actually send understandable modbus requests. https://www.modbusdriver.com/diagslave.html

2.  `diagslave -p 5020 -m enc` Start on port 5020 and with RTU over TCP  (`-m enc        Encapsulated Modbus RTU over TCP`)
3.  Configure your gateway to send modbus request to this machine ip and port 5020  (make sure port is not firewalled)

This example output of 2 modbus requests
```bash
x@x:~/code/$ diagslave -p 5020 -m enc
diagslave 3.1 - FieldTalk(tm) Modbus(R) Diagnostic Slave Simulator
Copyright (c) 2002-2018 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: Encapsulated RTU over TCP
Slave configuration: address = -1, master activity t/o = 3.00
IP configuration: port = 5020, connection t/o = 60.00

Server started up successfully.
Listening to network (Ctrl-C to stop)
.
validateMasterIpAddr: accepting connection from 127.0.0.1
Slave   2: readHoldingRegisters from 256, 1 references
..
validateMasterIpAddr: accepting connection from 127.0.0.1
Slave   2: readHoldingRegisters from 256, 1 references
...
```

This validates it is possible to use regular modbus servers. 

After that it is worth starting to tweak https://github.com/aldas/modbus-tcp-client/blob/master/examples/example_response_server.php  example to accept modbus RTU requests and storing them somewhere


NB: if you can configure gateway to send Modbus TCP instead of Modbus RTU over TCP it would be (much) easier.

**Reply by aldas** [reply]
More technical way to reverse engineer that handshake protocol before modbus request would be to  capture network traffic between KIOT platform and your gateway.  

For that you need to be in same network as your gateway is and use  Wireshark to capture that traffic.  I have written some hints about that here https://github.com/aldas/modbus-tcp-client/wiki/Debugging-guide#capture-network-traffic-to-see-what-is-actually-sendreceived

**Reply by agalindo73** [reply]
Friend, I apologize for the delay in writing to you, but I have already carried out tests, but since the GSM modbus device does not have an ethernet port, I have no way to connect it within my network, so the diaslave test does not work for this case. However, I did a test from my computer and with a mobile phone with the modbus simulator app, the diagslave test if it worked. The internet provider that I have handles private dynamic ip, I came up with an idea I am going to make an instance with amazon AWS with a windows or linux machine with public ip and I will do a test with the device to see if it grabs the connection and returns with you
Thank you so much

**Reply by agalindo73** [reply]
Friend, here I am again, I had vacations and a lot of work, so I'm just going back to the project that we were talking about in this discussion, finally we made an Apache server and installed the reactphp to run your codes, before running your php codes, realize a test with the diagslave as I commented and if the test worked (I add photos of the diagslave test), after this we perform the test with its code example_response_server.php and it worked (I add photograph of the code test)

As we can see in the photo that the module sends the request to the server every 60 seconds, I am not sure if that frame that is sending me is the modbus frame? , I would like to know which of its codes could I see the more detailed information by registers, this module as we can see has only 6 analog inputs, 8 logical inputs, temperature / humidity input and input voltage, so I see that it should get information from at least 17 registers, and I don't know if within the codes that you put what is the one that would display the information of all the registers?

I am going step by step to understand this process with modbus, because later I want to send the information of those registers to a mysql database

Your help has helped me a lot, thank you

regards

Alonso Galin
![WhatsApp Image 2021-09-28 at 8 56 36 PM](https://user-images.githubusercontent.com/73866027/135202050-2311d885-e50a-4903-92b2-780c9a9cba80.jpeg)
![WhatsApp Image 2021-09-28 at 7 53 17 PM](https://user-images.githubusercontent.com/73866027/135202053-89edf865-9047-430c-9839-e5930b19f872.jpeg)
![61ULX-YjeNL _AC_SL1000_ (1)](https://user-images.githubusercontent.com/73866027/135202055-16111489-27f8-422f-86ff-c02ec6a1fcc5.jpg)
do

**Reply by aldas** [reply]
that HEX
```
3136303337334e4e34304b3035353534
726571
```
is in ASCII
```
160373NN40K05554
req
```

So. this is "login?" message of that proprietary protocol. Server needs to answer probably something specific to that. Currently it will be considered as invalid packet and nothing will be sent. This IF will always return https://github.com/aldas/modbus-tcp-client/blob/c39a0e51519c5ef76f6ff0f2fb8e36fb34de2935/examples/example_response_server.php#L46


What you really need to do is capture traffic to actual "kiiot platform" to understand login handshake protocol and then implement similar logic in this server

**Reply by agalindo73** [reply]
hello aldas
ok I already understood that the info is arriving in HEX that the module ID is arriving, that ID was provided to me by the king pigeon company when I contracted the use of their KIIOT platform. It is so that when my module logs into that platform of theirs. But in itself with this module I can make my own modbus server, it does not occupy ID (I can delete this ID and leave it blank)
There is no blockage in the communication, this communication must be transparent without problems (I add a conversation months ago with the king pigeon about this).
about what you tell me to intercept the communication and see how the handshake is done, I'm going to do it although I'm not very familiar with that, but this module has no locks so it should work without problems, I think that probably something should be modified in it code for sending the response, I attach the configuration window and in the part of Heartbeat ACK Message and login message strategy in this last option I think there is another option but until tomorrow I go to work I will confirm it
Thank you so muc
![113649130-beead100-9642-11eb-930e-6803e7707340 (2)](https://user-images.githubusercontent.com/73866027/135548342-3c821941-8b41-48da-9023-3536129d6aff.jpg)
![113804727-ff664f80-9713-11eb-84e9-7da39761ef56](https://user-images.githubusercontent.com/73866027/135548345-097a7fff-c1bf-4f60-9a0d-d2db71588c15.jpg)
h

**Reply by agalindo73** [reply]
I forgot to add the configuration menu only that the S272 module does not apply to the IOT options

![menuconfiguration](https://user-images.githubusercontent.com/73866027/135549101-0d3f55c3-23d7-4499-815c-54ee2d55bc47.jpg)

**Comment by agalindo73** [comment]
Hi aldas I did the test with the wireshark intercepting the traffic on my gateway and accessing the KIIOT platform I have the file, I would like you to see it, the IP of the KIIOT page is 47.244.37.235.

alonso
[pruebawireshark3.zip](https://github.com/aldas/modbus-tcp-client/files/7272677/pruebawireshark3.zip)
o

**Reply by aldas** [reply]
@agalindo73 could specify which ip is S272 and to which IP it should send modbus data in that trafficdump. 

It would maybe better if you add filter `ip.addr == xxx.xxx.xxx.xxx` before you start recording. Then power off s272 and power it on so wireshark would record login etc from the start. and make sure that at least some modbus data sending is recorded.

**Reply by agalindo73** [reply]
Hi aldas today I got the serial device USR-TCP232-410S and it came with some tools to monitor the output of the modbus via usb, here I send you the tests of what I captured I did the procedure you told me, the modbus equipment is configured right now to access modbusrtu.kprtu.com and port 4000
It was generating this data for me, which it then converted to ascii, can you tell me if this information is useful for you?

regards
[outputhex.zip](https://github.com/aldas/modbus-tcp-client/files/7319530/outputhex.zip)
[text-output.zip](https://github.com/aldas/modbus-tcp-client/files/7319531/text-output.zip)
![converthextoascii](https://user-images.githubusercontent.com/73866027/136732869-28df7222-ff2e-4ac2-a93b-70bd7fb5f4da.jpg)
![softtestUSRTCP232](https://user-images.githubusercontent.com/73866027/136732870-fd4bc641-0ede-4d75-a479-4a96ad4703aa.jpg)

**Reply by agalindo73** [reply]
here goes another test
hex
62 75 69 6C 64 3A 20 53 65 70 20 32 39 20 32 30 32 30 20 31 31 3A 30 31 3A 33 31 0D 0A67 73 6D 20 69 6E 69 74 2E 2E 2E 0D 0A67 73 6D 20 33 47 20 6D 6F 64 75 6C 65 0D 0A0D 0A 69 6D 65 69 3A 38 36 34 33 35 32 30 34 33 35 31 34 39 34 32 0A67 73 6D 20 69 6E 69 74 20 4F 4B 0D 0A 73 69 6E 67 61 6C 3A 33 31 0D 0A00 04 30 30 30 31 30 33 30 41 41 31 34 36 39 36 37 34 35 35 38 32 30 30 30 30 37 41 35 32 36 41 31 35 30 34 37 44 44 46 43 42 37 32 44 30 44 33 42 44 36 42 32 38 39 32 34 34 31 44 30 43 30 36 38 42 45 44 31 41 38 41 32 39 33 44 34 43 41 46 43 46 37 35 34 31 37 39 42 42 34 43 44 45 33 35 31 34 43 37 3639 31 33 33 34 34 44 39 46 44 44 36 31 33 36 43 38 31 41 36 36 44 37 43 42 42 41 35 39 36 43 44 37 35 30 34 30 44 46 46 37 42 32 35 43 31 37 39 33 45 44 31 41 30 41 36 41 42 39 30 44 44 46 43 39 36 36 32 45 31 41 36 43 44 37 35 30 32 30 45 42 45 44 46 34 32 45 31 37 37 33 43 39 36 30 42 42 38 36 32 32 44 39 32 43 32 36 37 35 33 38 31 42 36 44 35 36 39 33 43 31 36 38 42 33 35 41 38 43 39 36 41 33 43 39 37 36 30 44 30 35 0D 0A00 04 30 30 30 31 30 33 30 41 41 31 34 36 39 36 37 34 35 35 38 32 30 30 30 30 32 31 44 36 42 32 37 43 39 45 37 45 42 42 37 35 42 33 41 32 37 35 32 36 44 42 33 35 31 34 33 32 39 38 32 43 44 36 38 41 43 31 35 41 33 31 31 38 34 38 31 36 44 33 43 44 36 38 33 42 0D 0A67 73 6D 20 74 63 70 20 63 6F 6E 6E 65 63 74 20 30 20 2E 2E 2E 0D 0A67 73 6D 20 74 63 70 20 63 6F 6E 6E 65 63 74 20 30 20 4F 4B 0D 0A 67 70 72 73 20 74 78 20 31 36 3A 20 33 31 20 33 36 20 33 30 20 33 33 20 33 37 20 33 33 20 34 45 20 34 45 20 33 34 20 33 30 20 34 42 20 33 30 20 33 35 20 33 35 20 33 35 20 33 34 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 32 20 30 30 20 30 30 20 30 30 20 30 34 20 37 39 20 43 39 20 0D 0A 67 70 72 73 20 74 78 20 36 3A 20 30 31 20 30 32 20 30 31 20 30 30 20 41 31 20 38 38 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 30 20 30 30 20 30 34 20 46 31 20 43 39 20 0D 0A 67 70 72 73 20 74 78 20 31 33 3A 20 30 31 20 30 34 20 30 38 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 32 34 20 30 44 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 31 38 20 30 30 20 30 32 20 46 31 20 43 43 20 0D 0A 67 70 72 73 20 74 78 20 39 3A 20 30 31 20 30 34 20 30 34 20 30 39 20 32 34 20 32 33 20 41 30 20 41 31 20 35 42 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 45 20 30 30 20 30 31 20 35 30 20 30 39 20 0D 0A 67 70 72 73 20 74 78 20 37 3A 20 30 31 20 30 34 20 30 32 20 30 34 20 43 36 20 33 42 20 41 32 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 32 20 30 30 20 30 30 20 30 30 20 30 34 20 37 39 20 43 39 20 0D 0A 67 70 72 73 20 74 78 20 36 3A 20 30 31 20 30 32 20 30 31 20 30 30 20 41 31 20 38 38 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 30 20 30 30 20 30 34 20 46 31 20 43 39 20 0D 0A67 70 72 73 20 74 78 20 31 33 3A 20 30 31 20 30 34 20 30 38 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 32 34 20 30 44 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 31 38 20 30 30 20 30 32 20 46 31 20 43 43 20 0D 0A 67 70 72 73 20 74 78 20 39 3A 20 30 31 20 30 34 20 30 34 20 30 39 20 32 34 20 32 33 20 39 36 20 32 31 20 34 44 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 45 20 30 30 20 30 31 20 35 30 20 30 39 20 0D 0A 67 70 72 73 20 74 78 20 37 3A 20 30 31 20 30 34 20 30 32 20 30 34 20 43 36 20 33 42 20 41 32 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 32 20 30 30 20 30 30 20 30 30 20 30 34 20 37 39 20 43 39 20 0D 0A 67 70 72 73 20 74 78 20 36 3A 20 30 31 20 30 32 20 30 31 20 30 30 20 41 31 20 38 38 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 30 20 30 30 20 30 34 20 46 31 20 43 39 20 0D 0A 67 70 72 73 20 74 78 20 31 33 3A 20 30 31 20 30 34 20 30 38 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 32 34 20 30 44 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 31 38 20 30 30 20 30 32 20 46 31 20 43 43 20 0D 0A 67 70 72 73 20 74 78 20 39 3A 20 30 31 20 30 34 20 30 34 20 30 39 20 32 34 20 32 33 20 41 30 20 41 31 20 35 42 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 45 20 30 30 20 30 31 20 35 30 20 30 39 20 0D 0A 67 70 72 73 20 74 78 20 37 3A 20 30 31 20 30 34 20 30 32 20 30 34 20 43 36 20 33 42 20 41 32 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 32 20 30 30 20 30 30 20 30 30 20 30 34 20 37 39 20 43 39 20 0D 0A 67 70 72 73 20 74 78 20 36 3A 20 30 31 20 30 32 20 30 31 20 30 30 20 41 31 20 38 38 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 30 20 30 30 20 30 34 20 46 31 20 43 39 20 0D 0A 67 70 72 73 20 74 78 20 31 33 3A 20 30 31 20 30 34 20 30 38 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 32 34 20 30 44 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 31 38 20 30 30 20 30 32 20 46 31 20 43 43 20 0D 0A67 70 72 73 20 74 78 20 39 3A 20 30 31 20 30 34 20 30 34 20 30 39 20 32 34 20 32 33 20 39 36 20 32 31 20 34 44 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 45 20 30 30 20 30 31 20 35 30 20 30 39 20 0D 0A67 70 72 73 20 74 78 20 37 3A 20 30 31 20 30 34 20 30 32 20 30 34 20 43 36 20 33 42 20 41 32 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 32 20 30 30 20 30 30 20 30 30 20 30 34 20 37 39 20 43 39 20 0D 0A 67 70 72 73 20 74 78 20 36 3A 20 30 31 20 30 32 20 30 31 20 30 30 20 41 31 20 38 38 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 30 20 30 30 20 30 34 20 46 31 20 43 39 20 0D 0A 67 70 72 73 20 74 78 20 31 33 3A 20 30 31 20 30 34 20 30 38 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 32 34 20 30 44 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 31 38 20 30 30 20 30 32 20 46 31 20 43 43 20 0D 0A 67 70 72 73 20 74 78 20 39 3A 20 30 31 20 30 34 20 30 34 20 30 39 20 32 34 20 32 33 20 42 45 20 32 31 20 35 33 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 45 20 30 30 20 30 31 20 35 30 20 30 39 20 0D 0A 67 70 72 73 20 74 78 20 37 3A 20 30 31 20 30 34 20 30 32 20 30 34 20 43 36 20 33 42 20 41 32 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 32 20 30 30 20 30 30 20 30 30 20 30 34 20 37 39 20 43 39 20 0D 0A 67 70 72 73 20 74 78 20 36 3A 20 30 31 20 30 32 20 30 31 20 30 30 20 41 31 20 38 38 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 30 20 30 30 20 30 34 20 46 31 20 43 39 20 0D 0A67 70 72 73 20 74 78 20 31 33 3A 20 30 31 20 30 34 20 30 38 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 30 30 20 32 34 20 30 44 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 31 38 20 30 30 20 30 32 20 46 31 20 43 43 20 0D 0A 67 70 72 73 20 74 78 20 39 3A 20 30 31 20 30 34 20 30 34 20 30 39 20 32 34 20 32 33 20 41 41 20 32 31 20 35 43 20 0D 0A67 70 72 73 20 72 78 20 38 3A 20 30 31 20 30 34 20 30 30 20 30 45 20 30 30 20 30 31 20 35 30 20 30 39 20 0D 0A 67 70 72 73 20 74 78 20 37 3A 20 30 31 20 30 34 20 30 32 20 30 34 20 43 36 20 33 42 20 41 32 20 0D 0A

ascii
build: Sep 29 2020 11:01:31
gsm init...
gsm 3G module

imei:864352043514942
gsm init OK
singal:31
0001030AA1469674558200007A526A15047DDFCB72D0D3BD6B2892441D0C068BED1A8A293D4CAFCF754179BB4CDE3514C76913344D9FDD6136C81A66D7CBBA596CD75040DFF7B25C1793ED1A0A6AB90DDFC9662E1A6CD75020EBEDF42E1773C960BB8622D92C2675381B6D5693C168B35A8C96A3C9760D05
0001030AA14696745582000021D6B27C9E7EBB75B3A27526DB351432982CD68AC15A31184816D3CD683B
gsm tcp connect 0 ...
gsm tcp connect 0 OK
gprs tx 16: 31 36 30 33 37 33 4E 4E 34 30 4B 30 35 35 35 34 
gprs rx 8: 01 02 00 00 00 04 79 C9 
gprs tx 6: 01 02 01 00 A1 88 
gprs rx 8: 01 04 00 00 00 04 F1 C9 
gprs tx 13: 01 04 08 00 00 00 00 00 00 00 00 24 0D 
gprs rx 8: 01 04 00 18 00 02 F1 CC 
gprs tx 9: 01 04 04 09 24 23 A0 A1 5B 
gprs rx 8: 01 04 00 0E 00 01 50 09 
gprs tx 7: 01 04 02 04 C6 3B A2 
gprs rx 8: 01 02 00 00 00 04 79 C9 
gprs tx 6: 01 02 01 00 A1 88 
gprs rx 8: 01 04 00 00 00 04 F1 C9 
gprs tx 13: 01 04 08 00 00 00 00 00 00 00 00 24 0D 
gprs rx 8: 01 04 00 18 00 02 F1 CC 
gprs tx 9: 01 04 04 09 24 23 96 21 4D 
gprs rx 8: 01 04 00 0E 00 01 50 09 
gprs tx 7: 01 04 02 04 C6 3B A2 
gprs rx 8: 01 02 00 00 00 04 79 C9 
gprs tx 6: 01 02 01 00 A1 88 
gprs rx 8: 01 04 00 00 00 04 F1 C9 
gprs tx 13: 01 04 08 00 00 00 00 00 00 00 00 24 0D 
gprs rx 8: 01 04 00 18 00 02 F1 CC 
gprs tx 9: 01 04 04 09 24 23 A0 A1 5B 
gprs rx 8: 01 04 00 0E 00 01 50 09 
gprs tx 7: 01 04 02 04 C6 3B A2 
gprs rx 8: 01 02 00 00 00 04 79 C9 
gprs tx 6: 01 02 01 00 A1 88 
gprs rx 8: 01 04 00 00 00 04 F1 C9 
gprs tx 13: 01 04 08 00 00 00 00 00 00 00 00 24 0D 
gprs rx 8: 01 04 00 18 00 02 F1 CC 
gprs tx 9: 01 04 04 09 24 23 96 21 4D 
gprs rx 8: 01 04 00 0E 00 01 50 09 
gprs tx 7: 01 04 02 04 C6 3B A2 
gprs rx 8: 01 02 00 00 00 04 79 C9 
gprs tx 6: 01 02 01 00 A1 88 
gprs rx 8: 01 04 00 00 00 04 F1 C9 
gprs tx 13: 01 04 08 00 00 00 00 00 00 00 00 24 0D 
gprs rx 8: 01 04 00 18 00 02 F1 CC 
gprs tx 9: 01 04 04 09 24 23 BE 21 53 
gprs rx 8: 01 04 00 0E 00 01 50 09 
gprs tx 7: 01 04 02 04 C6 3B A2 
gprs rx 8: 01 02 00 00 00 04 79 C9 
gprs tx 6: 01 02 01 00 A1 88 
gprs rx 8: 01 04 00 00 00 04 F1 C9 
gprs tx 13: 01 04 08 00 00 00 00 00 00 00 00 24 0D 
gprs rx 8: 01 04 00 18 00 02 F1 CC 
gprs tx 9: 01 04 04 09 24 23 AA 21 5C 
gprs rx 8: 01 04 00 0E 00 01 50 09 
gprs tx 7: 01 04 02 04 C6 3B A2

**Reply by agalindo73** [reply]
Aldas I have been trying to understand the modbus operation, as you know I do not master it but I downloaded a king pigeon manual of a newer module than mine S273 to S275, however the logical operation is the same you can see in the last pages As is the sequence, to all this I have my module configured, the mapping register is as follows: the 4 digital outputs (deactivated), the digital inputs (DI0 - DI3 activated, DI4 - DI7 deactivated) and the analog inputs (AI0 - AI1 activated, AI2 - AI5 deactivated)

According to the last email I sent, I already saw the sequence that according to my configuration the king pigeon platform sends the response to the module so that it can send the information, I attach a sheet where I mark with yellow the sequences that I would like your scripts example_response_server.php would respond with these sequences, but I don't know how to do it, I see that in the part of your scripts in lines 87 - 104 it is the part where it defines this sequence. I would like to know if you could help me how to put these sequences
Aldas I have been trying to understand the modbus operation, as you know I do not master it but I downloaded a king pigeon manual of a newer module than mine S273 to S275, however the logical operation is the same you can see in the last pages As is the sequence, to all this I have my module configured, the mapping register is as follows: the 4 digital outputs (deactivated), the digital inputs (DI0 - DI3 activated, DI4 - DI7 deactivated) and the analog inputs (AI0 - AI1 activated, AI2 - AI5 deactivated)

According to the last email I sent, I already saw the sequence that according to my configuration the king pigeon platform sends the response to the module so that it can send the information, I attach a sheet where I mark with yellow the sequences that I would like your scripts example_response_server.php would respond with these sequences, but I don't know how to do it, I see that in the part of your scripts in lines 87 - 104 it is the part where it defines this sequence. I would like to know if you could help me how to put these sequences

my project only every 60 seconds the module sends data to the modbus server, and the data that it would send would be the states
DI0
DI1
DI2
DI3
AIN0
AIN1
External voltage
Temperature
Humidity

regards

![serverresponse](https://user-images.githubusercontent.com/73866027/137600612-07b06aec-ec4c-4c6c-8b12-40e0d4fa3b69.png)
[S273_S275_Cellular_IoT_Modbus_RTU_User_Manual_V1.2.zip](https://github.com/aldas/modbus-tcp-client/files/7358184/S273_S275_Cellular_IoT_Modbus_RTU_User_Manual_V1.2.zip)
ds

**Comment by agalindo73** [comment]
Aldas I think we are a bit confused, this equipment is a GSM module and does not have an ethernet port, the SIM that the mobile phone service provider gives me here in Mexico they do not offer static IP service, the internet service I have is IP private dynamics, so this complicates a lot to carry out the test that you ask me, I will have to buy a device I attach the photograph to you and connect my king pigeon module through the RS485 port to the RS485 to Ethernet serial equipment.
I am going to order it and as far as I get and I will do the tests, on the other hand, I think your PHP scripts example_response_server.php is working fine, but as it lacks to send the response "res" to the GSM client module so that it sends the data from the registers, only you can identify this, in the manual it says that once the request message ("req") is received the GSM module has 6 seconds to respond to the module (Heartbeat ACK message) and I see that with your php scripts if it receives the Heartbeat message ("req") but does not send the response ("res
![WhatsApp Image 2021-10-03 at 8 13 20 AM](https://user-images.githubusercontent.com/73866027/135777813-66642a1c-545f-476f-a9d6-e3540b585f4e.jpeg)
![135549101-0d3f55c3-23d7-4499-815c-54ee2d55bc47](https://user-images.githubusercontent.com/73866027/135777816-fe23ba76-92dc-49b3-beef-d932f9f6fe16.jpg)
")
